### ISSUE TITLE
[BUG]: Segfault on RHEL/CentOS when binding code that imports shared library and STL queue

### ISSUE THREAD

**Issue by CharlesB2** [pybind/pybind11#3386]
### Required prerequisites

- [X] Make sure you've read the [documentation](https://pybind11.readthedocs.io). Your issue may be addressed there.
- [x] Search the [issue tracker](https://github.com/pybind/pybind11/issues) and [Discussions](https:/pybind/pybind11/discussions) to verify that this hasn't already been reported. +1 or comment there if it has.
- [X] Consider asking first in the [Gitter chat room](https://gitter.im/pybind/Lobby) or in a [Discussion](https:/pybind/pybind11/discussions/new).

### Problem description

I met a crash on a project with CentOS, that I could isolate with the following minimal conditions:

If a shared library has class with a `std::queue` member, using this class in a pybind11 module will lead to a segmentation fault.

So yes, all these conditions must be met:
* If the library is static, or code is directly included in the pybind module, no crash.
* If you run it on Ubuntu, no crash.
* And if you replace the `std::queue` with a `std::vector`, no crash.

(Note that it's probably not a duplicate of #3380)

Adding a complete PR with the below code in a few moments

#### crashing.cpp
```cpp

#include "crashing.h"

crashing::crashing()
= default;
```

#### crashing.h
```cpp
#include <queue>

class crashing{
public:
    crashing();

private:
    std::queue<int> queue_;
};
```
#### pybind.cpp
```cpp
#include "crashing.h"
#include <pybind11/pybind11.h>

void crash()
{
    crashing crash;
}

PYBIND11_MODULE(crash_shared, m) {
    m.def("crash", &crash);
}
```
#### CMakeLists.txt

```cmake
cmake_minimum_required(VERSION 3.12)
project(crash)

add_library(libcrash SHARED crashing.cpp)
target_compile_features(libcrash PRIVATE cxx_std_14)

pybind11_add_module(crash_shared pybind.cpp)
target_link_libraries(crash_shared PUBLIC
        $<TARGET_LINKER_FILE_DIR:libcrash>/$<TARGET_LINKER_FILE_NAME:libcrash>)
add_dependencies(crash_shared libcrash)
```

#### pybind.py
```py
import crash_shared
crash_shared.crash()
```

**Comment by CharlesB2** [comment]
After checking out https://github.com/pybind/pybind11/discussions/3308, it looks like I have the same issue, using compilers from conda-forge solves my issue.

This has been very hard for me to figure out, so I'm wondering is there a way at some point to detect such mismatch and add a warning, e.g. at CMake compilation?

**Comment by CharlesB2** [comment]
Thanks for the suggestion!

The better fix I found for this is to build using a recent manylinux image for building, and do an [auditwheel](https://github.com/pypa/auditwheel) on the resulting wheel (I'm not sure if the  auditwheel is needed, it might need triage. Using [cibuildwheel](cibuildwheel.readthedocs.io/) does nicely both manylinux build and auditwheel.

**Comment by ka-bo** [comment]
Thats also nice workaround. 

However,  that sounds like the compiler / _glibc_ version in _CentOS 8_ might by the problem. Hence, you cannot build a _pybind11_ extension module in _CentOS 8_ reliably, at least when you utilize the _pybind11_ _cmake_ tools and the default optimization they do.

**Comment by ka-bo** [comment]
Maybe the issue should remain open? I would not consider this solved or closed, as long as building an extension module via _pybind11 cmake_ tools is not working reliably in _CentOS 8_.

**Comment by CharlesB2** [comment]
You're right, let's leave it open and wait for maintainers' opinion about it

**Comment by jhill2** [comment]
I ran into this issue too with a very similar implementation as you've shown here. I was using pybind11_add_module in CMake to create a shared object on a RHEL 7, gcc 7.2.1, CMake 3.14, pybind11 2.4 instance.

I upgraded to RHEL 8, gcc 8.4.1, CMake 3.20, pybind11 2.8 and started to get a seg fault when I imported the module in Python. I fixed it using the 'NO_EXTRAS' flag that @ka-bo suggested.

A note on the versions, when I downgraded the gcc version in the updated software stack to gcc 7.2.1 it worked again without the 'NO_EXTRAS' flag so I agree that it's something with the compiler.

We are not using std::queue in the shared library we're linking against that is causing the seg fault. There must be more to it than just that.
